### ISSUE TITLE
Performance bottleneck when running Ruby

### ISSUE THREAD

**Issue by KernelErr** [WasmEdge/WasmEdge#1016]
There is a pull request for adding WASI support of Ruby. Here is a compiled module(Original, not optimized):

[miniruby.zip](https://github.com/WasmEdge/WasmEdge/files/7886204/miniruby.zip)

However, while running the following script to calculate PI:

```ruby
num = 4.0
pi = 0
plus = true

den = 1
while den < 1000000
  if plus
    pi = pi + num/den
    plus = false
  else
    pi = pi - num/den
    plus = true
  end
  den = den + 2
end

puts "PI = #{pi}"
```

WasmEdge seems slower than other runtimes, even after AOT.

![image](https://user-images.githubusercontent.com/45716019/149872681-a290fd9f-068b-4489-bbe4-5b4c7afaf9a5.png)

I am wondering if there was a performance bottleneck. There is a Valgrind callgrind profile:

[callgrind.zip](https://github.com/WasmEdge/WasmEdge/files/7886220/callgrind.zip)

![valgrind](https://user-images.githubusercontent.com/45716019/149872877-4d37b5da-3f0e-49b8-8098-28a263b5829a.png)

**Comment by q82419** [comment]
The 11% of execution time is stuck with the unnecessary loading for the WASM instructions in AOT mode.
This part will be fixed in #1274 .

**Comment by q82419** [comment]
The 35% of execution time is stuck with the unnecessary checking when getting the instances on calling frames.
This part will be fixed in #1337 .
